File: src/cpp/smart_fill.cpp
#include "smart_fill.hpp"
#include <queue>
#include <cmath>

namespace AnimStudio {

std::vector<uint8_t> SmartFill::floodFillMask(
    const uint8_t* imageData,
    int width,
    int height,
    int startX,
    int startY,
    ColorRGBA targetColor,
    int tolerance
) {
    std::vector<uint8_t> mask(width * height, 0);

    if (startX < 0 || startX >= width || startY < 0 || startY >= height) {
        return mask;
    }

    auto colorMatch = [&](int px, int py) -> bool {
        int idx = (py * width + px) * 4;
        int dr = std::abs(static_cast<int>(imageData[idx]) - targetColor.r);
        int dg = std::abs(static_cast<int>(imageData[idx + 1]) - targetColor.g);
        int db = std::abs(static_cast<int>(imageData[idx + 2]) - targetColor.b);
        int da = std::abs(static_cast<int>(imageData[idx + 3]) - targetColor.a);
        return (dr + dg + db + da) <= (tolerance * 4);
    };

    std::queue<std::pair<int, int>> q;
    q.push({startX, startY});
    mask[startY * width + startX] = 255;

    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};

    while (!q.empty()) {
        auto [cx, cy] = q.front();
        q.pop();

        for (int i = 0; i < 4; ++i) {
            int nx = cx + dx[i];
            int ny = cy + dy[i];

            if (nx >= 0 && nx < width && ny >= 0 && ny < height) {
                int maskIdx = ny * width + nx;
                if (mask[maskIdx] == 0 && colorMatch(nx, ny)) {
                    mask[maskIdx] = 255;
                    q.push({nx, ny});
                }
            }
        }
    }

    return mask;
}

} // namespace AnimStudio

File: animstudio/smart_fill.py
"""
Smart Fill and Flood-Fill Engine for AnimStudio 7.
Computes region fills, encloses open vector paths, and generates fill polygon shapes.
"""

from typing import List, Tuple, Set
from .types import Point, VectorObject


def hex_to_rgb(hex_str: str) -> Tuple[int, int, int]:
    hex_str = hex_str.lstrip('#')
    if len(hex_str) == 6:
        return tuple(int(hex_str[i:i+2], 16) for i in (0, 2, 4))
    return (0, 0, 0)


def smart_flood_fill(image_data: List[List[Tuple[int, int, int, int]]], start_x: int, start_y: int, fill_color: Tuple[int, int, int, int], tolerance: int = 32) -> Set[Tuple[int, int]]:
    """Performs 4-way BFS flood fill algorithm on a 2D pixel buffer with color tolerance."""
    height = len(image_data)
    if height == 0:
        return set()
    width = len(image_data[0])

    if start_x < 0 or start_x >= width or start_y < 0 or start_y >= height:
        return set()

    target_color = image_data[start_y][start_x]
    
    def color_match(c1, c2):
        return sum(abs(a - b) for a, b in zip(c1, c2)) <= tolerance * 4

    visited = set()
    queue = [(start_x, start_y)]
    filled_pixels = set()

    while queue:
        x, y = queue.pop(0)
        if (x, y) in visited:
            continue
        visited.add((x, y))

        if color_match(image_data[y][x], target_color):
            filled_pixels.add((x, y))
            for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nx, ny = x + dx, y + dy
                if 0 <= nx < width and 0 <= ny < height and (nx, ny) not in visited:
                    queue.append((nx, ny))

    return filled_pixels


def close_path_for_fill(points: List[Point]) -> List[Point]:
    """Ensures a vector path forms a closed polygon for fill rendering."""
    if not points:
        return []
    res = list(points)
    if res[0].x != res[-1].x or res[0].y != res[-1].y:
        res.append(Point(x=res[0].x, y=res[0].y, z=res[0].z, color=res[0].color, thickness=res[0].thickness))
    return res
